### ISSUE TITLE
UNET3+ with DS and CGM

### ISSUE THREAD

**Issue by rtsmapping-yili** [hamidriasat/UNet-3-Plus#4]
Hi,

May I ask which part of the code should I change to make unet3+ with DS and CGM trainable? I got INVALID_ARGUMENT:  required broadcastable shapes error when trying to train with it.

Specifically, I am only using your code to initialize the model, hopefully, I want it to be trained with the input shape of (None, None,8) for binary segmentation. Please let me know if you want more details.

Best
Yili

**Comment by hamidriasat** [comment]
Hi,

To support CGM training, I believe you need to update two items.

The first is in [data_generator.py](https://github.com/hamidriasat/UNet-3-Plus/blob/69bcb36169191f57846029e6d2592c750ae8b7d6/data_generator.py#L161) file. In data generator instead of returning only images and mask you need to return something like below
`[batch_images, batch_images, batch_images, batch_images, batch_images, batch_classification], batch_masks` 

`batch_images` are already there, you need to add classification input.  Here  `batch_classification` will be of shape batch_size x 1. It's a classification task that shows if required objects exist in an image or not. 0 if it doesn't and 1 if it does exist.


Second change is in the [loss function](https://github.com/hamidriasat/UNet-3-Plus/blob/69bcb36169191f57846029e6d2592c750ae8b7d6/train.py#L95), You can update it like this
`loss = [unet3p_hybrid_loss, unet3p_hybrid_loss, unet3p_hybrid_loss, unet3p_hybrid_loss, unet3p_hybrid_loss, tf.keras.losses.BinaryCrossentropy(from_logits=False)]`

Just for your understanding, we are passing separate input and loss functions for each model output. You can see [here](https://github.com/hamidriasat/UNet-3-Plus/blob/69bcb36169191f57846029e6d2592c750ae8b7d6/models/unet3plus_deep_supervision_cgm.py#L152) that model is returning six objects.

Last, during CGM training it may indicate that some gradients are not backpropagating warnings. This will be due to CGM so neglect it.

**Comment by mx5nabcd** [comment]
@hamidriasat 
> Last, during CGM training it may indicate that some gradients are not backpropagating warnings. This will be due to CGM so neglect it.

I'm using your source code thankfully. I have a question.

WARNING:tensorflow:Gradients do not exist for variables ['conv2d_10/kernel:0', 'conv2d_10/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_10/kernel:0', 'conv2d_10/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?

I looked up Conv2d_10, which is related to the warning, and it seems to be related to CGM.
Are these warnings about what you mentioned? Can I ignore it?

**Comment by hamid-cc** [comment]
@mx5nabcd 
Yes these are the ones I was referring, you can ignore these. they are because of CGM.

**Comment by mx5nabcd** [comment]
> @mx5nabcd Yes these are the ones I was referring, you can ignore these. they are because of CGM.

@hamidriasat 
Hello, I wonder why this warning occurs Do you know why?
